File: Leetcode100/leetcode100_python/SerializeandDeserializeBinaryTree3.py
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        if not root:
            return "X"
        left = "(" + self.serialize(root.left) + ")"
        right = "(" + self.serialize(root.right) + ")"
        return left + str(root.val) + right

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        if data == 'X':
            return None
        i = 0
        count = 0
        while i < len(data):
            if data[i] == '(':
                count += 1
            elif data[i] == ')':
                count -= 1
            if count == 0:
                break
            i += 1
        left = self.deserialize(data[1 : i])
        num_begin = i + 1
        i += 1
        while data[i] != "(":
            i += 1
        root = TreeNode(int(data[num_begin : i]))
        right = self.deserialize(data[i + 1: len(data) - 1])
        root.left, root.right = left, right
        return root

def printTree(root):
    if not root:
        return
    print(root.val)
    printTree(root.left)
    printTree(root.right)

if __name__ == "__main__":
    nums = [int(num) for num in input().split()]
    root = TreeNode(nums[0])
    queue = [root]
    for i in range(1, len(nums), 2):
        cur = queue.pop(0)
        if nums[i] != -1:
            cur.left = TreeNode(nums[i])
            queue.append(cur.left)
        if nums[i + 1] != -1:
            cur.right = TreeNode(nums[i + 1])
            queue.append(cur.right)
    
    codec = Codec()
    data = codec.serialize(root)
    print(data)
    root = codec.deserialize(data)
    printTree(root)

File: Leetcode100/leetcode100/src/main/java/com/jiang/leetcode100/serialize_and_deserialize_binary_tree/SerializeandDeserializeBinaryTree3.java
package main.java.com.jiang.leetcode100.serialize_and_deserialize_binary_tree;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.Scanner;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) {this.val = val;}
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Codec3 {

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        if(root == null) return "";
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        String result = "";
        while(!queue.isEmpty()) {

            TreeNode cur = queue.poll();
            if(cur == null) result += "#,";
            else {
                result += cur.val + ",";
                queue.add(cur.left);
                queue.add(cur.right);
            }
        }
        return result;
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        if(data.isEmpty()) return null;

        String[] nodes = data.split(",");
        TreeNode root = new TreeNode(Integer.parseInt(nodes[0]));
        Queue<TreeNode> queue = new ArrayDeque<>();
        queue.add(root);

        for(int i = 1; i < nodes.length; i += 2) {
            TreeNode cur = queue.poll();
            if(nodes[i].equals("#")) cur.left = null;
            else {
                cur.left = new TreeNode(Integer.parseInt(nodes[i]));
                queue.add(cur.left);
            }
            if(nodes[i + 1].equals("#")) cur.right = null;
            else {
                cur.right = new TreeNode(Integer.parseInt(nodes[i + 1]));
                queue.add(cur.right);
            }
        }
        return root;
    }
}

public class SerializeandDeserializeBinaryTree3 {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String str = scanner.nextLine();
        scanner.close();
        String[] nums_str = str.split(" ");
        int[] nums = new  int[nums_str.length];
        for(int i = 0; i < nums_str.length; i++) {
            nums[i] = Integer.parseInt(nums_str[i]);
        }

        TreeNode root = new TreeNode(nums[0], null, null);
        Queue<TreeNode> queue = new ArrayDeque<>();
        queue.add(root);

        for(int i = 1; i < nums.length; i += 2) {
            TreeNode cur = queue.poll();
            if(nums[i] != -1) {
                cur.left = new TreeNode(nums[i], null, null);
                queue.add(cur.left);
            }
            if(nums[i + 1] != -1) {
                cur.right = new TreeNode(nums[i + 1], null, null);
                queue.add(cur.right);
            }
        }

        Codec3 codec = new Codec3();
        String code = codec.serialize(root);
        System.out.println(code);
        TreeNode cur = codec.deserialize(code);
        printTree(cur);
    }

    private static void printTree(TreeNode root) {
        if(root == null) return;
        System.out.print(root.val + " ");
        printTree(root.left);
        printTree(root.right);
    }
}